go/libraries/doltcore/doltdb/commit_hooks.go

```diff
@@ -16,42 +16,91 @@ package doltdb
 
 import (
 	"context"
+	"fmt"
 	"io"
 	"sync"
 	"time"
 
-	"github.com/dolthub/go-mysql-server/sql"
-
 	"github.com/dolthub/dolt/go/libraries/doltcore/ref"
 	"github.com/dolthub/dolt/go/store/datas"
 	"github.com/dolthub/dolt/go/store/datas/pull"
 	"github.com/dolthub/dolt/go/store/hash"
 	"github.com/dolthub/dolt/go/store/types"
+	"github.com/dolthub/go-mysql-server/sql"
 )
 
 type PushOnWriteHook struct {
 	destDB datas.Database
 	tmpDir string
 	out    io.Writer
+	fmt    *types.NomsBinFormat
 }
 
 var _ CommitHook = (*PushOnWriteHook)(nil)
 
 // NewPushOnWriteHook creates a ReplicateHook, parameterizaed by the backup database
 // and a local tempfile for pushing
 func NewPushOnWriteHook(destDB *DoltDB, tmpDir string) *PushOnWriteHook {
-	return &PushOnWriteHook{destDB: destDB.db, tmpDir: tmpDir}
+	return &PushOnWriteHook{
+		destDB: destDB.db,
+		tmpDir: tmpDir,
+		fmt:    destDB.Format(),
+	}
 }
 
 // Execute implements CommitHook, replicates head updates to the destDb field
 func (ph *PushOnWriteHook) Execute(ctx context.Context, ds datas.Dataset, db datas.Database) error {
-	return pushDataset(ctx, ph.destDB, db, ph.tmpDir, ds)
+	// TODO: this code and pushDataset are largely duplicated from doltDb.PullChunks.
+	//  Clean it up, and preferably make more db stores capable of using the puller interface
+	if datas.CanUsePuller(db) && datas.CanUsePuller(ph.destDB) {
+		return pushDatasetWithPuller(ctx, ph.destDB, db, ph.tmpDir, ds)
+	}
+
+	return ph.pushDataset(ctx, ds, db)
+}
+
+func (ph *PushOnWriteHook) pushDataset(ctx context.Context, ds datas.Dataset, db datas.Database) error {
+	targetHash, ok := ds.MaybeHeadAddr()
+	if !ok {
+		return fmt.Errorf("dataset empty")
+	}
+
+	addr, ok := ds.MaybeHeadAddr()
+	if !ok {
+		_, err := ph.destDB.Delete(ctx, ds)
+		return err
+	}
+
+	rf, err := ref.Parse(ds.ID())
+	if err != nil {
+		return err
+	}
+
+	srcCS := datas.ChunkStoreFromDatabase(db)
+	destCS := datas.ChunkStoreFromDatabase(ph.destDB)
+	waf := types.WalkAddrsForNBF(ph.fmt)
+
+	err = pull.Pull(ctx, srcCS, destCS, waf, targetHash, nil)
+	if err != nil {
+		return err
+	}
+
+	ds, err = ph.destDB.GetDataset(ctx, rf.String())
+	if err != nil {
+		return err
+	}
+
+	_, err = ph.destDB.SetHead(ctx, ds, addr)
```

Comment:
Without this call, the pull.Pull above succeeds but results in a manifest with a 0 value root, and I don't get why